quickwit-storage/src/split.rs

```diff
@@ -0,0 +1,379 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::io::{self, Cursor, ErrorKind, SeekFrom, Write};
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+
+use async_trait::async_trait;
+use futures::StreamExt;
+use rusoto_core::ByteStream;
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+use tokio_util::io::ReaderStream;
+
+use crate::{BundleStorageFileOffsets, PutPayload};
+
+trait DataSource: tokio::io::AsyncRead + Send + Sync + Unpin {}
+impl<T: tokio::io::AsyncRead + Send + Sync + Unpin> DataSource for T {}
+
+/// SplitStreamer to Stream the split via the PutPayload.
+/// Bundles together multiple files into a single file
+/// with some metadata
+#[derive(Clone)]
+pub struct SplitStreamer {
+    /// Total split size
+    pub split_size: usize,
+    /// Files and their target range in the split bundle.
+    pub files_and_offsets: Vec<(PathBuf, Range<usize>)>,
+    /// bytes range of the footer (hotcache + bundle metadata)
+    pub footer_range: Range<usize>,
+    /// data of the footer, containng
+    /// [bundle metadata, bundle metadata len 8byte, hotcache, hotcache len 8 bytes]
+    pub footer_bytes: Vec<u8>,
+    /// The end offset of the data provider [File, .. File, SplitFooter]
+    pub block_offsets: Vec<Range<usize>>,
+}
+
+impl SplitStreamer {
+    fn get_start_and_end_index(&self, range: Range<usize>) -> io::Result<(usize, usize)> {
+        get_start_and_end_index(&self.block_offsets, range)
```

Comment:
Generally I prefer free functions over methods, because 
1. you can easily reuse them elsewhere
2. the required data is clear when reading and not just "self"